REST Data with Panache: support of entity event listeners

Sgitario opened the pull request:
REST Data with Panache supports the subscription to the following entity events: 

* Before/After save events
* Before/After update events
* Before/After delete events

To suscribe to an entity event, you need to provide a bean that implements the interface `EntityEventListener`, for example:

```java
@ApplicationScoped
public class PeopleRestDataResourceMethodListener implements RestDataResourceMethodListener<Person> {
    @Override
    public void onBeforeAdd(Person person) {
        System.out.println("Before Save Person: " + person.name);
    }
}
```

When using the REST Data Reactive with Panache extension, you will receive the Uni instance of the entity in the after methods:

```java
@ApplicationScoped
public class PeopleRestDataResourceMethodListener implements RestDataResourceMethodListener<Person> {
    @Override
    public void onAfterAdd(Uni<Person> uni) {
        uni.subscribe().with(person -> {
            System.out.println("After Save Person: " + person.name);
        });
    }
}
```

Fix https://github.com/quarkusio/quarkus/issues/29095

geoand commented:
Seems reasonable to me.

@yrodiere do you also think this makes sense?

yrodiere commented:
> @yrodiere do you also think this makes sense?

While it might seem reasonable at first glance, there's already a very similar feature available: JPA entity listeners. See https://github.com/quarkusio/quarkus/commit/0852b104d7341445af3cf8f41fa7be4e2758fd9d#diff-7ed082417d8730c92d0dcf5c59e547878a5c298311afbd529f1a5c9591f150cc for a few tests.

I'm struggling to see what this brings, beyond duplicating an already available, standard feature?

Note it's also more limited:

* It supports fewer event types than JPA entity listeners
* The listener won't be called for, e.g. cascaded persists or updates, while JPA entity listeners would.

If it's about getting to manipulate a Uni, then I'd suggest improving Hibernate Reactive so that it supports Uni in JPA entity listeners?

Sgitario commented:
PR updated according to the comments in https://github.com/quarkusio/quarkus/issues/29095
cc @geoand

geoand commented:
I can't really say I know much about the subject in hand, so I'll let @yrodiere and @gsmet decide.

yrodiere commented on extensions/panache/hibernate-orm-rest-data-panache/deployment/src/main/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/HibernateOrmPanacheRestProcessor.java:
```diff
@@ -134,6 +154,14 @@ private void validateResource(IndexView index, ClassInfo classInfo) {
+    private List<ClassInfo> getListenersByEntityType(List<ResourceMethodListenerBuildItem> resourceMethodListeners,
+            String entityTypeName) {
+        return resourceMethodListeners.stream()
+                .filter(e -> e.getEntityType().name().toString().equals(entityTypeName))
```
Just a heads-up: this doesn't seem to handle inheritance.

Sgitario replied:
Thanks for spotting! I've already addressed this.

yrodiere commented on extensions/panache/hibernate-reactive-rest-data-panache/deployment/src/test/resources/application.properties:
```diff
@@ -2,6 +2,6 @@ quarkus.datasource.db-kind=postgresql
 quarkus.datasource.password=hibernate_orm_test
 quarkus.datasource.reactive=true
-quarkus.datasource.reactive.url=${postgres.reactive.url}
+quarkus.datasource.reactive.url=vertx-reactive:postgresql://localhost:5431/hibernate_orm_test
```
Was this intentional?

Sgitario replied:
Ops

yrodiere reviewed (approved):
So I dug up a bit the Git history, and I must say I was wrong: judging from his past contributions to this extensions, @geoand knows *more* about it than I do :)

Anyway: @gytis was the original creator of this extension, and @Sgitario seems to have contributed the most since then. So... I'll trust him on the functional side :)

On the technical side, I must say I'm doubtful about the benefits of all the renamings, considering the noise it brings to this PR overall, but I guess there are reasons. Other than that I just spotted a few things; I'll let you see if you want to address them, @Sgitario. Thanks!

geoand commented:
I'll try and have a look later on today or tomorrow

Sgitario commented:
> So I dug up a bit the Git history, and I must say I was wrong: judging from his past contributions to this extensions, @geoand knows _more_ about it than I do :)
> 
> Anyway: @gytis was the original creator of this extension, and @Sgitario seems to have contributed the most since then. So... I'll trust him on the functional side :)
> 
> On the technical side, I must say I'm doubtful about the benefits of all the renamings, considering the noise it brings to this PR overall, but I guess there are reasons. Other than that I just spotted a few things; I'll let you see if you want to address them, @Sgitario. Thanks!

Yes, I had refactored the entities used in the tests to reuse the same ones, but it's true that it adds lots of noise in the pull request, so I've reverted these changes to the model and will do it in a separate ticket. Hopefully, now the pull request will be easier to review. 

PR updated.

yrodiere commented on extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/ResourceMethodListenerUtils.java:
```diff
@@ -0,0 +1,49 @@
+
+    public static List<ClassInfo> getListenersByEntityType(List<ResourceMethodListenerBuildItem> resourceMethodListeners,
+            String entityTypeName) {
+        Class<?> entityClass = loadClass(entityTypeName);
```
I loading classes in the deployment is generally frowned upon, though I forgot why exactly (performance? problems with the deployment classloader?).

Anyway, can you use the Jandex index instead? You should be able to retrieve a `ClassInfo` from `CombinedIndexBuildItem`, then retrieve all superclasses and superinterfaces from that `ClassInfo`. We do that in several places already, e.g. https://github.com/quarkusio/quarkus/blob/a23639025a26b3f4073efd64a5188191ca0067f6/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java#L451-L456

Sgitario replied:
Ok ok, I'm not sure if there is a performance benefit, but for the deployment extensions, the Index is generally used, so I've already changed it. Many thanks for checking!

geoand commented on docs/src/main/asciidoc/rest-data-panache.adoc:
```diff
@@ -354,6 +354,42 @@ It applies to the paged resources only and is a number starting with 1. Default
+public class PeopleRestDataResourceMethodListener implements RestDataResourceMethodListener<Person> {
+    @Override
+    public void onAfterAdd(Uni<Person> uni) {
+        uni.subscribe().with(person -> {
```
I would like to know more about why this is `Uni` here. 
Furthermore, is this same `Uni` that is returned by HR? If so, what happens when something like RR subscribed to the same `Uni`?

geoand replied:
To be honest, this does not seem terribly user friendly, why would we not provide users the entity itself?

geoand replied:
Yeah I understand the reasoning. But it's not very user friend really - there is almost never a good reason to force users to subscribe to a `Uni`

Sgitario committed:
REST Data with Panache: support of entity event listeners

REST Data with Panache supports the subscription to the following entity events:

* Before/After save events
* Before/After update events

To suscribe to an entity event, you need to provide a bean that implements the interface `EntityEventListener`, for example:

```java
@ApplicationScoped
public class PeopleEventListener implements EntityEventListener<Person> {
    @Override
    public void onBeforeSave(Person person) {
        System.out.println("Before Save Person: " + person.name);
    }
}
```

When using the REST Data Reactive with Panache extension, you will receive the Uni instance of the entity in the after methods:

```java
@ApplicationScoped
public class PeopleEventListener implements EntityEventListener<Person> {
    @Override
    public void onAfterSave(Uni<Person> uni) {
        uni.subscribe().with(person -> {
            System.out.println("After Save Person: " + person.name);
        });
    }
}
```

Fix https://github.com/quarkusio/quarkus/issues/29095

geoand reviewed (approved):
Great!

geoand commented:
Actually I have one more question:

Currently the implementation subscribes to the `Uni` for HR - do we really want that or is it better to instead just use something like `onItem`?
The reason I am asking is because the use of this will cause the `Uni` to be evaluated no matter what - whereas otherwise if the user just created the `Uni` but didn't return it to Quarkus, nothing would have happened.

I am not sure what the proper behavior is here.

Sgitario commented:
> Actually I have one more question:
> 
> Currently the implementation subscribes to the `Uni` for HR - do we really want that or is it better to instead just use something like `onItem`? The reason I am asking is because the use of this will cause the `Uni` to be evaluated no matter what - whereas otherwise if the user just created the `Uni` but didn't return it to Quarkus, nothing would have happened.
> 
> I am not sure what the proper behavior is here.

When I implemented this feature, I tried first using onItem and didn't work (the listeners were not invoked), but I can try again. Will keep you posted.

geoand commented:
> I tried first using onItem and didn't work (the listeners were not invoked)

Yeah, if the Uni is not subscribed to by Quarkus (or the user), then nothing will get executed.

I am wondering what behavior we really want though... I can see arguments for either way

Sgitario commented:
> > I tried first using onItem and didn't work (the listeners were not invoked)
> 
> Yeah, if the Uni is not subscribed to by Quarkus (or the user), then nothing will get executed.
> 
> I am wondering what behavior we really want though... I can see arguments for either way

I just confirmed that using `onItem().invoke()` will not execute the after listeners. So, I think our only way is to keep using the `suscribe().with()` approach. 
Also, take into account that this will only be used if there is at least one `after` listener implemented by user, not always, so there won't be performance penalty here.

geoand commented:
Okay, fine from side :)

Sgitario commented:
Merging this pull request as all reviewers are ok with these changes.
